### PR TITLE
Usage Based Billing

### PR DESCRIPTION
**Pull request by bhelx**
This feature that is not yet enabled on all sites.

This introduces some new resources and updates to current resources to
support the usage based billing (metered billing) feature.

https://docs.recurly.com/docs/metered-billing

### Using

```ruby
require 'date'
require 'securerandom'

# Let's create a new plan
plan_code = SecureRandom.uuid
plan = Recurly::Plan.create!(
  plan_code: plan_code,
  name: 'My Plan',
  unit_amount_in_cents: { 'USD' => 10_00, 'EUR' => 8_00 },
  setup_fee_in_cents: { 'USD' => 60_00, 'EUR' => 45_00 },
  plan_interval_length: 1,
  plan_interval_unit: 'months',
  tax_exempt: false
)

# Create the plan add_on
add_on = plan.add_ons.create!(
  add_on_code: 'marketing_emails',
  name: 'Marketing Emails',
  unit_amount_in_cents: 5,
  add_on_type: :usage,
  optional: true,
  usage_type: :price,
  measured_unit_id: 18783, # You need to create this measured unit in the dashboard
)

# This should be an account with a good billing info
account = Recurly::Account.find 'usageaccount'

# Create the subscription add on
sub_add_on = Recurly::SubscriptionAddOn.new('marketing_emails')
sub_add_on.quantity = 1

# create the subscription with the addon
subscription = Recurly::Subscription.create!(
  plan_code: plan_code,
  subscription_add_ons: [sub_add_on],
  account: {
    account_code: account.account_code
  }
)

# we have this variable already, but finding it for demonstration purposes
sub_add_on = subscription.subscription_add_ons.detect do |add_on|
  add_on.add_on_code == 'marketing_emails'
end

sleep 1

usage1 = sub_add_on.usage.build(
  amount: 100,
  merchant_tag: "Batch of 100 marketing emails sent by customer",
  recording_timestamp: DateTime.now,
  usage_timestamp: DateTime.now
)

usage1.save!

usage2 = sub_add_on.usage.build

usage2.amount = 50
usage2.merchant_tag = "Batch of 50 marketing emails sent by customer"
usage2.recording_timestamp = DateTime.now
usage2.usage_timestamp = DateTime.now

usage2.save!

# You can create instead of build
usage3 = sub_add_on.usage.create!(
  amount: 200,
  merchant_tag: "Batch of 200 marketing emails sent by customer",
  recording_timestamp: DateTime.now,
  usage_timestamp: DateTime.now
)

# Our usage search options
search_options = {
  billing_status: :all,   # (:all || :unbilled || :billed)
  datetime_type: :usage,  # (:usage || :recording)
#  start_datetime: NOW,    # greater than or equal to this date
#  end_datetime: NOW,      # less than this date
}

# create the pager
# takes per_page and stuff
# see Pager class
pager = sub_add_on.usage.paginate(search_options)

# iterate through our 3 usages
pager.find_each do |usage|
  puts usage.inspect
end
```

### CODE REVIEW

**Comment by pmorton** [comment]
Well this is exciting. I am hoping that this is for billing for things like storage on a monthly basis!

**Comment by bhelx** [comment]
@pmorton yes! we have some new features coming down the pipe: https://docs.recurly.com/docs/metered-billing

It's not enabled by default yet but you can contact support if it's something you want to try out.

File: Gemfile [hunk 1]
```diff
@@ -10,4 +10,7 @@ group :development do
   gem 'rubysl',                    :platforms => :rbx
   gem 'rubysl-resolv', '~> 2.0',   :platforms => :rbx
   gem 'racc',                      :platforms => :rbx
+
+  gem 'webmock', '~> 1.7'
+  gem 'pry'
```
Comment (bhelx) [inline Gemfile]: I'm going to do a follow up PR to remove the gemfile and put stuff in the gemspec

File: lib/recurly.rb [hunk 2]
```diff
@@ -15,11 +15,13 @@ module Recurly
   require 'recurly/invoice'
   require 'recurly/js'
   require 'recurly/money'
+  require 'recurly/measured_unit'
```
Comment (csmb) [inline lib/recurly.rb]: Bugs me that this require list is halfway organized alphabetically.

**Comment by csmb** [comment]
Do you want to include a Changelog updated in this PR?

File: lib/recurly.rb (hunk 2 above)
Comment (bhelx) [inline lib/recurly.rb]: Yes me too, I'd rather do a change like that in another commit or another PR though.

**Comment by bhelx** [comment]
@csmb No, I'm going to update the changelog for the release. There are a lot of other things that need to go into it.

**Comment by csmb** [comment]
Looks good to me +1
